torch/utils/data/datapipes/iter/routeddecoder.py

```diff
@@ -1,38 +1,47 @@
-from torch.utils.data import IterDataPipe
+import re
+
+from io import BufferedIOBase
+from typing import Any, Callable, Iterable, Iterator, Sized, Tuple
+
+from torch.utils.data import IterDataPipe, functional_datapipe
 from torch.utils.data.datapipes.utils.decoder import (
     Decoder,
     basichandlers as decoder_basichandlers,
     imagehandler as decoder_imagehandler)
 
-from typing import Iterable, Iterator, Union, List, Tuple, Any, Callable
-from io import BufferedIOBase
 
+# Extract extension from pathname
+def _default_key_fn(pathname):
+    return re.sub(r".*[.]", "", pathname)
+
+
+@functional_datapipe('decode')
 class RoutedDecoderIterDataPipe(IterDataPipe[Tuple[str, Any]]):
     r""" :class:`RoutedDecoderIterDataPipe`.
 
-    Iterable datapipe to decode binary streams from input iterables,
-    yield pathname and decoded binary stream in a tuple.
+    Iterable datapipe to decode binary streams from input DataPipe, yield pathname
+    and decoded data in a tuple.
     args:
         datapipe: Iterable datapipe that provides pathname and binary stream in tuples
-        handlers: user defined decoder handlers, if None, basic and image decoder handlers will be set as default
-        length: a nominal length of the datapipe
+        handlers: Optional user defined decoder handlers. If None, basic and image decoder
+            handlers will be set as default. If multiple handles are provided, the priority
+            order follows the order of handlers (the first handler has the top priority)
+        key_fn: Optional function for decoder to extract key from pathname to dispatch handlers.
+            Default is set to extract extension from pathname
     """
 
-    def __init__(
-            self,
-            datapipe : Iterable[Tuple[str, BufferedIOBase]],
-            *,
-            handlers : Union[None, List[Callable]] = None,
-            length: int = -1):
+    def __init__(self,
+                 datapipe: Iterable[Tuple[str, BufferedIOBase]],
+                 *handlers: Callable,
+                 key_fn: Callable = _default_key_fn) -> None:
         super().__init__()
-        self.datapipe : Iterable[Tuple[str, BufferedIOBase]] = datapipe
-        if handlers:
-            self.decoder = Decoder(handlers)
+        self.datapipe: Iterable[Tuple[str, BufferedIOBase]] = datapipe
+        if len(handlers) > 0:
+            self.decoder = Decoder(handlers, key_fn)
```

Comment:
Instead of having two branches, maybe just simply set a default value if nothing is passed:

```suggestion
        if not handlers:
            handlers = [decoder_basichandlers, decoder_imagehandler('torch')]
        self.decoder = Decoder(handlers, key_fn)
```

(this should replace the full `if` / `else`, but I can't add a suggestion for this)